networkx/algorithms/approximation/broadcasting.py

```diff
@@ -0,0 +1,139 @@
+    if not nx.is_tree(G):
+        NetworkXError("Your graph is not a tree")
+    # step 0
+    if G.number_of_nodes() == 2:
+        return 1, set(G.nodes())
+    elif G.number_of_nodes() == 1:
+        return 0, set(G.nodes())
+
+    # step 1
+    U = {node for node, deg in G.degree if deg == 1}
+    values = {n: 0 for n in U}
+    T = G.copy()
+    T.remove_nodes_from(U)
+
+    # step 2
+    W = {node for node, deg in T.degree if deg == 1}
+    values.update((w, G.degree[w] - 1) for w in W)
+
+    # step 3
+    while T.number_of_nodes() >= 2:
+        # step 4
+        w = min(W, key=lambda n: values[n])
+        v = next(T.neighbors(w))
+
+        # step 5
+        U.add(w)
+        W.remove(w)
+        T.remove_node(w)
+
+        # step 6
+        if T.degree(v) == 1:
+            # update t(v)
+            values.update({v: _get_max_broadcast_value(G, U, v, values)})
+            W.add(v)
+
+    # step 7
+    v = nx.utils.arbitrary_element(T)
+    b_T = _get_max_broadcast_value(G, U, v, values)
+    return b_T, _get_broadcast_centers(G, v, values, b_T)
+
+
+@not_implemented_for("directed")
+@not_implemented_for("multigraph")
+def tree_broadcast_time(G, node=None):
+    """Return the Broadcast Time of the tree G.
+
+    Broadcasting is an information dissemination problem in which one vertex in a graph, called the originator,
+    must distribute a message to all other vertices by placing a series of calls along the edges of the graph.
+    Once informed, other vertices aid the originator in distributing the message.
+    The broadcasting must be completed as quickly as possible subject to the following constraints:
+    - Each call requires one unit of time.
+    - A vertex can only participate in one call per unit of time.
+    - Each call only involves two adjacent vertices: a sender and a receiver.
+    The minimum broadcast time of a vertex is defined as the minimum amount of time required to complete
+    broadcasting starting from the originator.
+    The broadcast time of a graph is defined as the maximum broadcast time from all its vertices.
+
+    Parameters
+    ----------
+    G : NetworkX graph
+        Undirected graph
+        The graph should be an undirected tree
+    node: int
+        index of starting vertex. If none,
+        the algorithm returns the broadcast
+        time of the tree.
+    Returns
+    -------
+    BT : int
+        Broadcast Time of a vertex in a tree
+
+    Raises
+    ------
+    NetworkXNotImplemented
+        If the graph is directed or is a multigraph.
+
+    References
+    ----------
+    .. [1] Slater, P.J., Cockayne, E.J., Hedetniemi, S.T,
+       Information dissemination in trees. SIAM J.Comput. 10(4), 692–701 (1981)
+    """
+    b_T, b_C = tree_broadcast_center(G)
+    if node is not None:
+        return b_T + min(nx.shortest_path_length(G, node, u) for u in b_C)
+    else:
+        return b_T + max(min(nx.shortest_path_length(G, v, u) for u in b_C) for v in G)
```

Comment:
Naturally, now I need to check the min distances of all vertices in the graph $G$ and take the largest one. I believe there could be a way to optimize this, I tried using eccentricity but there were some edge cases that it couldn't detect, please let me know if you think of something clever!

Reply:
I tried this implementation, but it seemed to return some errors for the tests I wrote. 
The thing is for the broadcast time of the graph, I need to find the furthest vertex from the broadcast centers (any in particular, in this case). 
As for the broadcast time of a particular vertex, my implementation seems correct since I am taking the minimum distance from the source to all the broadcast centers. 
Could you elaborate a bit more on your optimization? I feel like I am slightly unsure why it's correct (I understand that there are way less shortest paths calculations though). 
Maybe we could go backwards, and you could point out where my implementation is not efficient?

Reply:
Also P.S. I made some changes to your code block since there was a small error.
Maybe this is not what you intended? I checked the docs and ``nx.shortest_path_length`` can receive only the graph as a parameter.
```python
dist_center = dict.fromkeys(G, len(G))
for u, dd in nx.shortest_path_length(G): # changes on this line as u was undefined
    for v, dist in dd.items():
        if dist < dist_center[v]:
            dist_center[v] = dist
return b_T + max(dist_center.values())
```

Reply:
Sorry -- I have the code inconsistent with itself there... rats.  I was trying to get a single call to `nx.shortest_path_length` instead of calling it or every `u, v` pair.  So to find the broadcast time of the tree, we should loop over the nodes in b_C, calculate the shortest path from that node to all nodes, and then store that in a dict "dist_from_center" if it is shorter than all previous distances to that node.

Hopefully this is closer to actually running:
```python
dist_from_center = dict.fromkeys(G, len(G))
for u in b_C:
    for v, dist in nx.shortest_path_length(G, u).items():
        if dist < dist_from_center[v]:
            dist_from_center[v] = dist
return b_T + max(dist_center.values())
```
This is similar to what you have done with:
```python
max(min(nx.shortest_path_length(G, v, u) for u in b_C) for v in G)
```
but it switches the order of the loops  so you only need a shortest_path algorithm traversal for each node in b_C instead of for each pair in b_C \cross G.  To switch the order of the loops, we have to store the dist_from_center values outside of the loop because we don't do all the distances for node `v` inside a single `min` statement. We are comparing all the distances for node `v` spread out with all the other v's in between (i.e. our outer loop is over b_C).  I hope that is more understandable.  It's like the difference between "putting all the v-distances in a box for node v, then finding the min, then switching to the next node v" vs "making len(G) boxes -- one for each node v, then finding the distances to all nodes from a node u in b_C, taking each v-distance and looking at the v-box value and replacing it if the distance is smaller. That way, the box holds the min-so-far of the v-distances to nodes in b_C (each box holds one value). Repeat for all u in b_C and each box will hold the minimum distance to a node in b_C." But the second way only does |b_C| shortest paths to all nodes instead of |G|*|b_C| shortest_paths to a single target within b_C.

Hopefully that is more clear. Ask again if something isn't clear.

Reply:
Thank you for the explanation. I understand the idea now. Basically we now calculate the distance from every vertex to the centers but in reversed order (starting from the centers). And storing the distances allows us to cut an order of |G| shortest paths calculations.
This is quite a neat optimization, thanks for that.